Concurrent executor failover to serial queue

bmeike opened the pull request:
There are two important changes, here:
1) The parallel executor will never use the last MIN_CAPACITY (currently 64) slots of a bounded executor queue.  That leave a fair amount of space for serial queues (2 slots per queue) and other application stuff.
2) When the parallel executor exhausts execution resources (availCapacity - MIN_CAPACITY <= 0) it fails over to using an unbounded serial queue.  This means that the failure mode for, e.g., conflict revolvers that permanently block all execution, will be an eventual OOM.  Before the OOM warnings will be logged to help identify the actual problem.  Unfortunately, though, because the queue is unbounded, there will be no helpful execution queue dump.
All tests passing on all platforms

pasin commented on lib/src/shared/main/java/com/couchbase/lite/internal/AbstractExecutionService.java:
```diff
@@ -75,33 +98,92 @@ public String toString() {
+                }
+
+                n = pendingTasks.size();
+                if (n <= 1) { executeTask(newTask); }
```
I'm not sure I understand everything here. Is there a case like below?
1. There are several tasks in the pendingTasks waiting to get executed.
2. There is a new task come in but at that time spaceAvailable() returns true as are some running tasks are done right before the new task was added.
3.  newTask created in line 154 will not get executed by line 161 as n > 1.

borrrden commented on lib/src/shared/main/java/com/couchbase/lite/internal/AbstractExecutionService.java:
```diff
@@ -112,136 +194,182 @@ public boolean stop(long timeout, @NonNull TimeUnit unit) {
+                    if (task == null) { return; }
+                    executeTask(task);
                 }
+                while (spaceAvailable());
```
What happens in the case that multiple tasks pile up into the pending task queue, then the first one executes and that pushed it back out of space again.  It seems like this method will exit without completing them all.  When does it come back around when space is available again?

pasin replied:
Creating two types (one will call finishTask() and the other will call scheduleNext() ) of tasks depending on the available space at the time could be a concern here that that the scheduleNext() might not get called. Maybe when each single task is done, go through the same code path that also tries to schedule the next task?

bmeike replied:
I'm not sure what "pushed out into space again" means.  The currently running task is always the task at the head of the queue.  It can't get lost until it is completed.
I don't see how it is possible that `scheduleNext` could not be called...  Say more?

borrrden commented on lib/src/shared/main/java/com/couchbase/lite/internal/AbstractExecutionService.java:
```diff
@@ -75,33 +98,92 @@ public String toString() {
+                if (spaceAvailable()) { newTask = new InstrumentedTask(task, this::finishTask); }
+                else {
+                    newTask = new InstrumentedTask(task, this::scheduleNext);
+                    pendingTasks.add(newTask);
```
Part 2 of the comment below.  Since `executeTask` only gets called when there is one or zero items in this queue, it's possibly game over if it reaches 2 and no new tasks will ever execute:

1. This logic is reached, and `scheduleNext` will get executed
2. More tasks come into the `pendingTasks` list from somewhere else
3. `scheduleNext` executes a few, but runs out of space on the executor before finishing, leaving at least 1 item in `pendingTasks`
4. `pendingTasks` receives a second item and no items are ever executed again

borrrden commented on lib/src/shared/main/java/com/couchbase/lite/internal/AbstractExecutionService.java:
```diff
@@ -75,33 +98,92 @@ public String toString() {
-                if (stopLatch == null) { stopLatch = new CountDownLatch(running); }
+                if (stopLatch == null) {
+                    pendingTasks.clear();
+                    stopLatch = new CountDownLatch(1);
```
Why was this changed to 1?  It seems like this would get over counted down.

pasin replied:
Because of the change in line 200 that will only count down the latch when all running tasks are done?

borrrden commented on lib/src/shared/main/java/com/couchbase/lite/internal/AbstractExecutionService.java:
```diff
@@ -112,136 +194,182 @@ public boolean stop(long timeout, @NonNull TimeUnit unit) {
+                needsRestart = false;
+            }
+            catch (RejectedExecutionException e) {
+                needsRestart = true;
```
What if `execute` is never called again?  Won't this executor just stop forever?

pasin replied:
Same question here.

bmeike replied:
This implementation attempts to restart the queue when client code submits a new task
It was the best I could come up with.  If you can think of something better that would be great.
Here is my analysis:
There are no execution resources.  Attempting to re-schedule the task is pointless and, essentially, a spin lock.  Attempting to reschedule the task after a delay might make some sense.  In the normal case, though -- a permanently stalled task --  it is also pointless.
I've thought about some mechanism that would throw the REE to the use when they attempt to submit to a stalled queue.  That is definitely appealing...

borrrden commented:
This is a bit hard to follow given all the formatting changes that also took place, but I gave it a shot.

bmeike commented:
@borrrden  Yes.  Sorry about that.  It is, essentially, a rewrite.  I suggest reading the new code, and not the changes.

bmeike commented:
I believe that there are 3 bugs here:
1) Pasin described a condition in which, if the serial queue in non-empty, a task can be lost
2) Pasin and Jim both pointed out that if an task from the serial queue is rejected, the queue is hung forever.  The implementation of the SerialExecutor addresses this; the implementation needs to be pulled over.
3) I noticed that when the serial queue dispatches multiple items (because there is space available in the underlying executor), each (instead of only one) of them will reschedule from the queue.

bmeike commented:
I believe I have addressed all of the issues.  I have not yet run tests, and I think it is worth adding a couple more.  I would like to get your eyes on the new code, though ASAP.

bmeike commented:
All tests pass.  I still think I will add a couple more.  Esp for queue restart.

pasin commented:
@bmeike Need help to understand the definition or functionality of the restartQueue(). From the code, the execute one task from the top of the pending queue and when it's done it will try to execute a lot of tasks in the pending queue until the space is not available. Is it by design to execute only one task from the pending queue first (why we don't try to execute all the tasks in the pending queue)?

pasin commented on lib/src/shared/main/java/com/couchbase/lite/internal/AbstractExecutionService.java:
```diff
@@ -75,33 +105,113 @@ public String toString() {
+                if (spaceAvailable()) {
+                    if (needsRestart) { restartQueue(); }
+
+                    executeTask(new InstrumentedTask(task, this::finishTask));
```
After restart the queue, do we need to check again if space is still available?

bmeike replied:
It seems reasonable, to me, not to do that.  If, between the time that I check `spaceAvailable`, and the time I try to execute the task, the underlying executor is overwhelmed, then executeTask will throw.  That is the contract, whether `restartQueue` is called, in between, or not.

pasin replied:
I could think about one edge case that could cause the single pending task not running.
1. Assuming that each task submitted is a long running task. A lot of tasks get submitted until the space is unavailable.
2. One more task get submitted while the space is unavailable. The task get added to `pendingTasks`.
3. The restartQueue() is executed but because the space is unavailable, the reject exception is called, which will be ignored. Because there is no more task to come so the restartQueue() will not get called again so the single pending task will not be run.

pasin replied:
Beside that edge case, I think the implementation would work.

bmeike replied:
I think that is not quite correct:  A task is never removed from the queue until it has been successfully submitted for execution, right?  I think that is always true.  If so, when the queue is restarted, by a subsequent call to `execute` the queue will be restarted and there will be another attempt to submit the head.

bmeike commented:
`RestartQueue` does exactly what its name implies. It tries to schedule the head of `pendingQueue` on the executor.  In order to "restart" the queue, that task must call `scheduleNext` on completion.  If the method succeeds, it clears `needsRestart`, because the queue has been restarted.  If it fails it sets `needsRestart` because there is no running task that will schedule the next task.

bmeike commented:
Added a test for serial queue restart.  It even found a heinous bug in the logging routine!
It is proving much harder to write a test for restarting the concurrent executor.

bmeike commented:
Added a test for concurrent queue restart.

borrrden commented on lib/src/shared/main/java/com/couchbase/lite/internal/AbstractExecutionService.java:
```diff
@@ -75,33 +105,113 @@ public String toString() {
+                pendingTasks.add(new InstrumentedTask(task));
+
+                pendingTaskCount = pendingTasks.size();
+                if (needsRestart || (pendingTaskCount == 1)) { restartQueue(); }
```
Is it a good idea to restart the queue when no space is available?  Or will there possibly be space available now after a few lines from when you checked before?

bmeike replied:
Yeah... 
1) the contract is that the serial queue is not controlled by `spaceAvailable`.  The concurrent executor can fill all but the last 62 slots: 64 - 2 (for the serial queue).  That needs documentation
2) `restartQueue` will attempt to restart the queue, but will fail, if the executor is truly swamped.  It that case, it will continue to be stalled...

bmeike commented:
I am not delighted with this solution.  It *does* mitigate the biggest problem we have had so far: A serial executor getting an REE.  With that in mind, I am going to merge it to Mercury, with a note that the design needs to be revisited.
